src/client/Microsoft.Identity.Client/PoP/PoPAuthenticationScheme.cs

```diff
@@ -0,0 +1,132 @@
+        private readonly Uri _requestUri;
+        private readonly IPoPCryptoProvider _popCryptoProvider;
+        private readonly byte[] _keyThumbprint;
+
+        /// <summary>
+        /// Creates POP tokens, i.e. tokens that are bound to an HTTP request and are digitally signed.
+        /// </summary>
+        /// <remarks>
+        /// Currently the signing credential algorithm is hard-coded to RSA with SHA256. Extensibility should be done
+        /// by integrating Wilson's SigningCredentials
+        /// </remarks>
+        public PoPAuthenticationScheme(Uri requestUri, HttpMethod httpMethod, IPoPCryptoProvider popCryptoProvider)
+        {
+            _httpMethod = httpMethod ?? throw new ArgumentNullException(nameof(httpMethod));
+            _requestUri = requestUri ?? throw new ArgumentNullException(nameof(requestUri));
+            _popCryptoProvider = popCryptoProvider ?? throw new ArgumentNullException(nameof(popCryptoProvider));
+
+            _keyThumbprint = ComputeRsaThumbprint(_popCryptoProvider.CannonicalPublicKeyJwk);
+            KeyId = Base64UrlHelpers.Encode(_keyThumbprint);
+        }
+
+        public string AuthorizationHeaderPrefix => PoPRequestParameters.PoPAuthHeaderPrefix;
+
+        /// <summary>
+        /// For PoP, we chose to use the base64(jwk_thumbprint)
+        /// </summary>
+        public string KeyId { get; }
+
+        public IDictionary<string, string> GetTokenRequestParams()
+        {
+            return new Dictionary<string, string>() {
+                { OAuth2Parameter.TokenType, PoPRequestParameters.PoPTokenType},
+                { PoPRequestParameters.RequestConfirmation, ComputeReqCnf()}
+            };
+        }
+
+        public string FormatAccessToken(MsalAccessTokenCacheItem atItem)
+        {
+            var header = new JObject
+            {
+                { JsonWebTokenConstants.ReservedHeaderParameters.Algorithm, JsonWebTokenConstants.Algorithms.RsaSha256 },
+                { JsonWebTokenConstants.ReservedHeaderParameters.KeyId, KeyId },
+                { JsonWebTokenConstants.ReservedHeaderParameters.Type, PoPRequestParameters.PoPTokenType}
+            };
+
+            JToken popAssertion = JToken.Parse(_popCryptoProvider.CannonicalPublicKeyJwk);
+
+            var payload = new JObject(
+                new JProperty(PoPClaimTypes.At, atItem.Secret),
+                new JProperty(PoPClaimTypes.HttpMethod, _httpMethod.ToString()),
+                new JProperty(PoPClaimTypes.Host, _requestUri.Host),
+                new JProperty(PoPClaimTypes.Path, _requestUri.AbsolutePath),
+                new JProperty(PoPClaimTypes.Nonce, CreateNonce()),
+                new JProperty(PoPClaimTypes.Cnf, new JObject(
+                    new JProperty(PoPClaimTypes.JWK, popAssertion))));
+
+            return CreateJWS(payload.ToString(Json.Formatting.None), header.ToString(Json.Formatting.None));
+        }
+
+        private static string CreateNonce()
```

Comment:
do we need to verify that the nonce is valid? the spec mentions 1-5 min lifetime.

Reply:
I believe this is the responsability of the verifier (i.e. Wilson). MSAL should simply generate a random number so that the same request cannot be played twice.

Reply:
got it. thanks.

---
In reply to: [352705636](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1525#discussion_r352705636) [](ancestors = 352705636)